### PR TITLE
 [AST] Properly adjust substitution type for inherited conformances.

### PR DESCRIPTION
**Pull request by DougGregor**
When we substitute into an inherited conformance, make sure that we
follow the superclass chain from the new conforming type up to the
matching superclass *before* doing the substitution.

Fixes rdar://problem/35632543. Also add a test for the related but already-fixed rdar://problem/35480728.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: lib/AST/ProtocolConformance.cpp [hunk 1]
```diff
@@ -979,7 +979,17 @@ ProtocolConformance::subst(Type substType,
       = cast<InheritedProtocolConformance>(this)->getInheritedConformance();
     ProtocolConformance *newBase;
     if (inheritedConformance->getType()->isSpecialized()) {
-      newBase = inheritedConformance->subst(substType, subs, conformances);
+      // Follow the substituted type up the superclass chain until we reach
+      // the underlying class type.
+      auto targetClass =
+        inheritedConformance->getType()->getClassOrBoundGenericClass();
+      auto superclassType = substType;
+      while (superclassType->getClassOrBoundGenericClass() != targetClass)
```
Comment (slavapestov) [inline lib/AST/ProtocolConformance.cpp]: This is just substType->getSuperclassForDecl(targetClass)

Comment (DougGregor) [reply lib/AST/ProtocolConformance.cpp]: Oh, of course it is! Thank you.

Comment (DougGregor) [reply lib/AST/ProtocolConformance.cpp]: I'm pulling the cleanup into my next PR.

**Pull request merged**